src/main/java/org/opensearch/neuralsearch/query/AgenticSearchQueryBuilder.java

```diff
@@ -0,0 +1,273 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.opensearch.neuralsearch.query;
+
+import lombok.AccessLevel;
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.Setter;
+
+import lombok.NoArgsConstructor;
+import lombok.experimental.Accessors;
+import lombok.extern.log4j.Log4j2;
+import org.apache.commons.lang.builder.EqualsBuilder;
+import org.apache.commons.lang.builder.HashCodeBuilder;
+import org.apache.lucene.search.Query;
+import org.opensearch.core.ParseField;
+import org.opensearch.core.action.ActionListener;
+import org.opensearch.core.common.ParsingException;
+import org.opensearch.core.common.io.stream.StreamInput;
+import org.opensearch.core.common.io.stream.StreamOutput;
+import org.opensearch.core.xcontent.XContentBuilder;
+import org.opensearch.core.xcontent.XContentParser;
+import org.opensearch.index.query.AbstractQueryBuilder;
+import org.opensearch.index.query.WithFieldName;
+import org.opensearch.index.query.QueryBuilder;
+import org.opensearch.index.query.QueryRewriteContext;
+import org.opensearch.index.query.QueryCoordinatorContext;
+import org.opensearch.index.query.QueryShardContext;
+import org.opensearch.neuralsearch.ml.MLCommonsClientAccessor;
+import org.opensearch.common.xcontent.XContentType;
+import org.opensearch.common.xcontent.LoggingDeprecationHandler;
+import org.opensearch.neuralsearch.util.NeuralSearchClusterUtil;
+import com.google.gson.Gson;
+
+import java.util.Objects;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.List;
+import java.util.Locale;
+import java.util.ArrayList;
+import java.util.concurrent.CompletableFuture;
+
+import java.io.IOException;
+
+import static org.opensearch.core.xcontent.XContentParserUtils.ensureExpectedToken;
+
+/**
+ * AgenticSearchQueryBuilder is responsible for agentic query type. It executes a root agent to extract the DSL query and parse it to innerQueryBuilder for parsing.
+ */
+@Accessors(chain = true, fluent = true)
+@NoArgsConstructor
+@AllArgsConstructor
+@Getter(AccessLevel.PACKAGE)
+@Setter(AccessLevel.PACKAGE)
+@Log4j2
+public final class AgenticSearchQueryBuilder extends AbstractQueryBuilder<AgenticSearchQueryBuilder> implements WithFieldName {
+
+    public static final String NAME = "agentic";
+    public static final ParseField QUERY_TEXT_FIELD = new ParseField("query_text");
+    public static final ParseField AGENT_ID = new ParseField("agent_id");
+    public static final ParseField QUERY_FIELDS = new ParseField("query_fields");
+    private static final Gson gson = new Gson();
```

Comment:
have you checked if there is already something in OpenSearch that does the same. I mean if we can minimize number of external dependencies that's good

Reply:
yes, core doesn't have gson